Feat: Display unread marker for messages that were received while offline

chrispader opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/44007
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

1. Log into Expensify with two accounts on web + a different device or incognito tab.
2. Open a 1:1 conversation between the two accounts
3. Disable the internet connection of the first Expensify session in the DevTools (Network -> Throttling -> Offline)
4. Send messages from the second account to the first account
5. Enable internet again in the first session (Network -> Throttling -> No throttling)
6. There should be a "New" (unread messages) marker above the first message that was sent while the user was offline  

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

The description in "Tests" is already covering offline behaviour.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

Same as in Tests.

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/cbff560d-71df-44b4-939c-0af793c70161

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

MacOS Chrome (Arc) screen recording attached.

chrispader committed:
feat: unread marker for messages while offline

chrispader committed:
Merge branch 'Expensify:main' into @chrispader/unread-marker-not-displayed

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

chrispader committed:
add comments

chrispader committed:
add comment and fix logic

chrispader commented:
This is ready for review! :)

cc @brunovjk @alexpensify

brunovjk commented:
@roryabraham @chrispader  Do you think we shouldwork on the marker dismissal here?  I can only dismiss if I enter another conversation and come back:

https://github.com/user-attachments/assets/0924b04d-36e3-4c4b-bf60-e7e5fd7d2919

brunovjk commented:
@roryabraham Can you confirm something please? Should this [new feat of ours ](https://github.com/Expensify/App/issues/44007#issuecomment-2310973989) only work on the web or other platforms as well? Thank you.

chrispader commented:
> @roryabraham @chrispader Do you think we should work on the marker dismissal here? I can only dismiss if I enter another conversation and come back:

If so, i think this should be handled in a separate issue. The same behavior is the case for receiving messages in a report when it's not active. (also the case on `main`).


https://github.com/user-attachments/assets/1e71f8da-2184-444e-8349-c84bc1ad6274

brunovjk commented:
> > Do you think we should work on the marker dismissal here? I can only dismiss if I enter another conversation and come back:
> 
> If so, i think this should be handled in a separate issue. 

I understand, I also agree that taking this to another issue would be more appropriate.

@chrispader I'm having problems with the marker, I'm "playing around" with two conversations and the behavior is still not consistent in my tests:

https://github.com/user-attachments/assets/222f7fe9-eec9-44f1-9142-b4e766053aae


I'll come back here tomorrow and test more, I'll clean everything up here, maybe it's something with my local. But, in your tests, did everything work correctly? Thank you.

chrispader commented:
> @chrispader I'm having problems with the marker, I'm "playing around" with two conversations and the behavior is still not consistent in my tests:

> I'll come back here tomorrow and test more, I'll clean everything up here, maybe it's something with my local. But, in your tests, did everything work correctly? Thank you.

One thing that makes me wonder: Why did the right user receive the message from the left one before throttling was disabled (before it went back online)? That's weird.

I can see that the users own messages are als recognized as "new messages". This shouldn't happen. I'm going to look into that!

Other than that, i had no problems with the unread marker while testing

brunovjk commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/278350d8-94e7-49c6-ba32-fa526e751065


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/82bf9720-5c45-475b-846b-f9be82e0a99e


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/876103d2-d204-41e8-9d51-a23b3a31bc59


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/ff0de1ea-61f6-43ae-98e3-8b06cac71cf7


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/443c3201-b97b-4fef-8de9-baf2df08b2e0


</details>

brunovjk commented:
@chrispader Indeed, sorry, ignore that report, it was an issue with my local. I tested the changes in all platforms, it seems fine so far. Did you have a chance to check _"I can see that the users own messages are als recognized as "new messages". This shouldn't happen. I'm going to look into that!"_? Thank you.

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

chrispader committed:
fix: don't show unread marker for own messages

chrispader commented:
> Did you have a chance to check _"I can see that the users own messages are als recognized as "new messages". This shouldn't happen. I'm going to look into that!"_? Thank you.

Just fixed the issue 👍

brunovjk reviewed (approved):
I retested it and it looks good to me:

https://github.com/user-attachments/assets/9077b1e6-ffdb-433d-83c7-a3314df8bd09


@chrispader  Do you know why the _"Performance Tests error"_ in the PR? Thanks.

chrispader commented:
> @chrispader Do you know why the "Performance Tests error" in the PR? Thanks.

No idea tbh :/ I don't think it's actually related to the changes. Had the same issue with a few other PRs

roryabraham commented:
> Do you think we should work on the marker dismissal here?

I don't think we should introduce this feature at this time.

roryabraham commented:
> Should [this new feat of ours](https://github.com/Expensify/App/issues/44007#issuecomment-2310973989) only work on the web or other platforms as well?

I don't see any reason why this would be web-only. So all platforms 🙂

tylerkaraszewski commented:
I've approved the PR but the performance tests are failing.

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

chrispader committed:
re-structure offline message detection logic

chrispader commented:
The `Reassure Performance Tests` still fail, but i guess this is just due to the extra check for offline messages? This will also take more processing for each message (up to the first "read" one)....

I'm not sure we can further optimize this change

brunovjk commented:
Hi @chrispader, sorry to bother you, but is there anything I can do/help you with to move forward with this PR? Thank you very much.

brunovjk commented:
Hi @chrispader, sorry to bother you, but is there anything I can do/help you with to move forward with this PR? Thank you very much.

cc: @tylerkaraszewski

tylerkaraszewski commented:
@mallenexpensify - do you know how we're supposed to handle these performance tests?

mallenexpensify commented:
I'm unfamiliar with but found this[ internal SO](https://stackoverflowteams.com/c/expensify/questions/17757) and the below deets. 

> We use Reassure to monitor our app's performance and quickly catch any slowdowns or issues. It's part of our CI setup, making it easy to see if things are getting faster or slower when we push new changes. If you need more details, check out the [README](https://github.com/Expensify/App/blob/main/tests/perf-test/README.md?plain=1) in the repository.
> 
> Reassure works on top of the existing React Testing Library. It runs performance tests twice – once on your current branch and once on the main branch – so we can spot any differences. The focus is on testing more real-world scenarios, like how full screens are performing, but it's also handy for checking the performance of specific functions.
> 
> If a test fails, take a look at which specific test it is. It could be because a component in the flow is rendering more times than expected, or maybe the performance of a function has dropped. Make sure to dig into the root cause. If it's a contributor's PR, ask them to explain why the render count increased compared to the main branch. Sometimes, this is unavoidable to achieve the desired outcome for the PR. If they can provide a clear explanation, it's okay to merge even if some tests fail.
> 
> This won't mess up other PRs! Your change will set a new baseline for future tests.
> 
> One thing to keep in mind: some changes might actually break the tests, and you'll need to investigate and fix them. For example, the mocking of certain components could need to be updated to match the changes you made to the component itself.

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

chrispader committed:
simplify offline message check

chrispader committed:
simplify code

chrispader committed:
WIP: improve unread marker check

chrispader committed:
fix: unread marker not shown

chrispader commented:
@brunovjk i updated the PR and simplified the checks a bit. I hope these changes fix the Reassure tests. Will check tmrw morning

brunovjk reviewed (approved):
Thanks  @chrispader I retested it and although the perf error appeared, the changes seem good to me:

https://github.com/user-attachments/assets/d1f0d9d7-7b7d-49db-ad65-771290bb6de6

I tried to give it a try but I still can't clearly identify why it's failing and how to fix it.

cc: @tylerkaraszewski

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

chrispader committed:
fix: move offline/online at logic into seperate hook and defer values

chrispader committed:
fix: use ref instead of state for lastOffline/lastOnline at

chrispader commented:
Okay i updated the code to use `useRef` instead of additional state. This brings down extra render count to only one extra render. The PR still works as expected.

> If a test fails, take a look at which specific test it is. It could be because a component in the flow is rendering more times than expected, or maybe the performance of a function has dropped. Make sure to dig into the root cause. If it's a contributor's PR, ask them to explain why the render count increased compared to the main branch. Sometimes, this is unavoidable to achieve the desired outcome for the PR. If they can provide a clear explanation, it's okay to merge even if some tests fail.

The extra logic depends on `useLocalize()` to be able to convert the `message.created` datetime to a localized Date object. So we won't be able to avoid the extra render unfortunately.

chrispader committed:
fix: move ReportAction util functions

chrispader commented:
I removed the `isMessageUnread` util function from `ReportActionsList`, because we have effectively the same util function already in `ReportActionsUtils`. I also moved the `wasMessagedCreatedWhileOffline` to `ReportActionsUtils`.

brunovjk commented:
Nice! Let me test it :D

chrispader commented:
Performance tests are also succeeding now somehow 👀

brunovjk commented:
I found an inconsistency, @chrispader can you confirm it? If both users are offline, and send a message to each other, the "green line" should appear for both, right?

https://github.com/user-attachments/assets/ab655944-aa4c-4c6a-95ca-b44304fd796c

brunovjk commented:
@chrispader I've tested it a few times, with different platforms, in different ways, and I can't reproduce a specific bug reliably, but the behavior of our "green line" is no longer consistent with the last tests. After lunch I'll come back to try to find what might be causing it. Did everything go well in your tests? Thank you.

tylerkaraszewski commented:
To be clear, we do *not* want to merge this yet? Can we add a [HOLD] to it in that case?

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

chrispader committed:
fix: fix and simplify offline message check

chrispader committed:
fix: wrong parameters

chrispader committed:
fix: remove dependency array eslint exception

chrispader committed:
fix: re-implement offline message logic

chrispader commented:
I've made some more findings based on further testing and on your observations in https://github.com/Expensify/App/pull/49480#issuecomment-2426840173

> I think I found an inconsistency, @chrispader can you confirm it? If both users are offline, and send a message to each other, the "green line" should appear for both, right?

The current implementation will show the green line for all messages a user/device receives that have a `created` timestamp later than the time User A went offline and earlier than when User A came back online.

When either user goes offline and sends a message, the Backend API call will be delayed until after the user comes back online. Therefore, the `created` timestamp of the report action will not be within the offline duration of the other user.

This also happens to both users, when you go offline on the host device... the `created` timestamps of messages from both users will be delayed.

The device/user that comes back online first - they will never/very very unlikely ever come online at exactly the same time- makes the Backend API call first and therefore only the other user will see the green line.

Scenario with timestamps (`X` = point in time)

1. You turn on the internet at `1` -> Both users/devices go offline
2. You send a message from User A at `2`
3. You send a message from User B at `3`
4. You turn the internet back online at `4`. One of the users comes back online first. Let's say User A is online first at `5`
6. The message from User A will be sent to the backend at `6`
7. User B comes back online at `7`
8. The message from User B will be sent to the backend at `8`
9. User B will receive a message from User A with `created` timestamp `6`. **User B will see the green line**, because User B was still offline at `6`.
10. User A will receive a message from User B with `created` timestamp `8`. **User A will not see the green line**, because User A was already back online at `8`.

chrispader commented:
## Solution

This logic has no control of whether the message has (actually) been created/sent while the sender was offline. Only messages received while the own device was offline will be "marked as unread" (green line).

Therefore, we have two options on how to handle this UI-wise.

### Approach 1

Stick with the current implementation and mark messages as unread, that have been received while the own user was offline. In case both users are offline, the user who is back online first will not see the "urnead marker". 

### Approach 2

Also mark messages as unread, that have been created while the sender was offline. For this, we will need to add an additional prop like `createdWhileOfflineAt` or `wasCreatedWhileOffline` to keep track of whether the message was created in an offline state.

I don't favor this approach though, since this would also mean, that all (online) users will receive "unread markers" (green lines) if some other user sends a message while offline. This might be very disturbing in groups especially.

Curious about your thoughts cc @brunovjk @tylerkaraszewski

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

brunovjk commented:
> Curious about your thoughts cc @brunovjk @tylerkaraszewski

Thanks for the detailed explanation, @chrispader. Apologies for any unnecessary extra work on this. I believe this is primarily an edge case and not reflective of typical user behavior, though I can't confirm that definitively (we'd need @tylerkaraszewski's input on that). From my perspective, we can proceed as is to avoid regressions.

brunovjk reviewed (approved):
Overall, the changes seem solid and well thought out. I agree with proceeding, in my tests everything works well:

<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/efdae891-79f9-4479-a2f6-0be499417b12


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/1102d3cc-4ba3-4010-8f26-255033c96a5f


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/61d6bce7-f667-4d67-979f-843d5a3fd758


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/4e3c1dc0-8854-4e1c-be68-2bdfd192de86


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/b74d51a9-0e1a-469f-8c86-365b0e86804d


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/d3df9580-25b9-4337-b73f-41d3084881fe


</details>

tylerkaraszewski commented:
I've read the scenario above and particularly point 9 is of interest:

> User A will receive a message from User B with created timestamp 8. User A will not see the green line, because User A was already back online at 8.

This says that user A receives a message *while online* and thus the green line is not updated.

I think that's fine, I don't think user A needs to care if user B was online or not when they initially tried to send it. When the message is actually able to be sent, user A is online, and the message arrives immediately.

I think that's fine. That lets us go with the current solution, correct?

brunovjk commented:
Great, we just need to resolve the conflicts to merge? Thanks.

chrispader committed:
Merge branch 'main' into @chrispader/unread-marker-not-displayed

chrispader commented:
Resolved conflicts!

tylerkaraszewski commented:
Can we address the failing unit test?

brunovjk commented:
I'm not sure what is causing, if is indeed our PR, I will investigate

chrispader committed:
fix: wrong condition in early return

chrispader commented:
> I'm not sure what is causing, if is indeed our PR, I will investigate and back here if I find something

Sorry my bad, while restructuring i the changes from `main` i unintentionally inverted the condition for one of the early-returns in the `shouldDisplayNewMarker` check.

Tests should now succeed

brunovjk commented:
Retesting

chrispader commented:
Just tested it and still works on my machine :)

brunovjk commented:
Indeed, everything looks very good to me:

https://github.com/user-attachments/assets/8a53c90f-73b2-48a3-b2ac-0fbdd37ea6bb

@tylerkaraszewski all yours.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

jjcoffee commented on src/libs/ReportActionsUtils.ts:
```diff
@@ -1806,10 +1807,38 @@ function getReportActionsLength() {
 
+function wasActionCreatedWhileOffline(action: ReportAction, isOffline: boolean, lastOfflineAt: Date | undefined, lastOnlineAt: Date | undefined, locale: Locale): boolean {
+    // The user was never online.
+    if (!lastOnlineAt) {
```
Coming from https://github.com/Expensify/App/issues/55990. We missed a case here, where if you received a message from a chat that's not open, then go offline and open that report, the new message marker won't show. This is because `lastOnlineAt` will be undefined here as it's stored in component state. More [details](https://github.com/Expensify/App/issues/55990#issuecomment-2633402983).
